contracts/svm/testing/tests/swap.rs

```diff
@@ -0,0 +1,568 @@
+        &mut svm,
+        &input_token.get_associated_token_address(&searcher.pubkey()),
+        input_token.get_amount_with_decimals(10f64),
+    ));
+    assert!(Token::token_balance_matches(
+        &mut svm,
+        &input_token.get_associated_token_address(&trader.pubkey()),
+        output_token.get_amount_with_decimals(0f64),
+    ));
+    assert!(Token::token_balance_matches(
+        &mut svm,
+        &input_token.get_associated_token_address(&get_express_relay_metadata_key()),
+        input_token.get_amount_with_decimals(0f64),
+    ));
+    assert!(Token::token_balance_matches(
+        &mut svm,
+        &input_token.get_associated_token_address(&express_relay_metadata.fee_receiver_relayer),
+        input_token.get_amount_with_decimals(0f64),
+    ));
+    assert!(Token::token_balance_matches(
+        &mut svm,
+        &router_input_ta,
+        input_token.get_amount_with_decimals(0f64),
+    ));
+
+    // output token balances
+    assert!(Token::token_balance_matches(
+        &mut svm,
+        &output_token.get_associated_token_address(&searcher.pubkey()),
+        output_token.get_amount_with_decimals(0f64),
+    ));
+    assert!(Token::token_balance_matches(
+        &mut svm,
+        &output_token.get_associated_token_address(&trader.pubkey()),
+        output_token.get_amount_with_decimals(10f64),
+    ));
+    assert!(Token::token_balance_matches(
+        &mut svm,
+        &router_output_ta,
+        output_token.get_amount_with_decimals(0f64),
+    ));
+    assert!(Token::token_balance_matches(
+        &mut svm,
+        &output_token.get_associated_token_address(&get_express_relay_metadata_key()),
+        output_token.get_amount_with_decimals(0f64),
+    ));
+    assert!(Token::token_balance_matches(
+        &mut svm,
+        &output_token.get_associated_token_address(&express_relay_metadata.fee_receiver_relayer),
+        output_token.get_amount_with_decimals(0f64),
+    ));
+
+    // input token fee
+    let swap_args = SwapArgs {
+        deadline:         i64::MAX,
+        amount_input:     input_token.get_amount_with_decimals(1f64),
+        amount_output:    output_token.get_amount_with_decimals(1f64),
+        referral_fee_bps: 3000,
+        fee_token:        FeeToken::Input,
+    };
+    let instructions = build_swap_instructions(
+        searcher.pubkey(),
+        trader.pubkey(),
+        None,
+        None,
+        router_input_ta,
+        express_relay_metadata.fee_receiver_relayer,
+        input_token.mint,
+        output_token.mint,
+        Some(input_token.token_program),
+        Some(output_token.token_program),
+        swap_args,
+    );
+    submit_transaction(&mut svm, &instructions, &searcher, &[&searcher, &trader]).unwrap();
+
+    // input token balances
+    assert!(Token::token_balance_matches(
+        &mut svm,
+        &input_token.get_associated_token_address(&searcher.pubkey()),
+        input_token.get_amount_with_decimals(9f64),
```

Comment:
you could maybe calculate these? but @danimhr may have different opinion on that

Reply:
no, there's no point in rewriting the math imo. It will be the same math as in the program.

Reply:
I could reuse `ExpressRelayMetadata::compute_swap_fee` if you want

Reply:
I could also use `ExpressRelayMetadata::compute_swap_fees` but overall I think this is better: I mentally calculated what it should be and bind it to those values.